ros/wavemap_ros/src/input_handler/pointcloud_undistorter.cc

```diff
@@ -0,0 +1,103 @@
+#include "wavemap_ros/input_handler/pointcloud_undistorter.h"
+
+#include <wavemap_ros_conversions/time_conversions.h>
+
+namespace wavemap {
+PointcloudUndistorter::Result PointcloudUndistorter::undistortPointcloud(
+    GenericStampedPointcloud& stamped_pointcloud,
+    PosedPointcloud<>& undistorted_pointcloud, const std::string& fixed_frame) {
+  // Get the time interval
+  const uint64_t start_time = stamped_pointcloud.getStartTime();
+  const uint64_t end_time = stamped_pointcloud.getEndTime();
+
+  // Calculate the step size for the undistortion transform buffer
+  const auto& points = stamped_pointcloud.getPoints();
+  constexpr int kNumTimeIntervals = 400;
+  constexpr int kNumTimeSteps = kNumTimeIntervals + 1;
+  const uint64_t step_size =
+      (points.back().time_offset - points.front().time_offset) /
+      (kNumTimeIntervals - 1);
+  const uint64_t buffer_start_time = start_time - step_size;
+  const uint64_t buffer_end_time = end_time + step_size;
+
+  // Make sure all transforms are available
+  if (!transformer_->isTransformAvailable(
+          fixed_frame, stamped_pointcloud.getSensorFrame(),
+          convert::nanoSecondsToRosTime(buffer_start_time))) {
+    return Result::kStartTimeNotInTfBuffer;
+  }
+  if (!transformer_->isTransformAvailable(
+          fixed_frame, stamped_pointcloud.getSensorFrame(),
+          convert::nanoSecondsToRosTime(buffer_end_time))) {
+    return Result::kEndTimeNotInTfBuffer;
+  }
+
+  // Buffer the transforms
+  std::vector<std::pair<uint64_t, Transformation3D>> timed_poses;
+  timed_poses.reserve(kNumTimeSteps);
+  for (unsigned int step_idx = 0u; step_idx < kNumTimeSteps; ++step_idx) {
+    auto& timed_pose = timed_poses.emplace_back();
+    timed_pose.first = start_time + step_idx * step_size;
+    if (!transformer_->lookupTransform(
+            fixed_frame, stamped_pointcloud.getSensorFrame(),
+            convert::nanoSecondsToRosTime(timed_pose.first),
+            timed_pose.second)) {
+      ROS_WARN_STREAM("Failed to buffer intermediate pose at time "
+                      << convert::nanoSecondsToRosTime(timed_pose.first)
+                      << ".");
+      return Result::kIntermediateTimeNotInTfBuffer;
+    }
+  }
+
+  // Motion undistort
+  // NOTE: The undistortion is done by transforming the points into a fixed
+  //       (inertial) frame using the sensor's pose at each point's timestamp.
+  const auto num_points = static_cast<int>(points.size());
+  Eigen::Matrix<FloatingPoint, 3, Eigen::Dynamic> t_W_points;
```

Comment:
This is also a preference thing, but I generally prefer to initialize all Eigen memory to 0 just to be safe I don't bleed garbage memory data later on, just in case I forget to assign values to something I expected to assign values to.

Reply:
All Eigen types in wavemap are default initialized to NaN (with the `EIGEN_INITIALIZE_MATRICES_BY_NAN` cmake flag), to avoid silent uninitialized memory errors. But maybe that's a bit too aggressive. I'll switch to `setZero(rows,cols)` instead of resize. Thanks.